Method setRequestMethodNonBulk() added to allow (non bulk) POST requests

silvershadowcc opened the pull request:
POST was already supported, because it's used for BulkTrack. This method should provide the option to choose POST in other scenarios (when using setTokenAuth for increased security for example).

silvershadowcc commented:
Would be able to solve #43

tsteur commented:
Thanks @silvershadowcc 

btw there will be likely still few more changes needed so that the token actually gets posted and not sent through a URL query parameter.

tsteur commented on MatomoTracker.php:
```diff
@@ -1477,6 +1477,19 @@ public function setRequestTimeout($timeout)
+	
+	/**
+     * Sets the request method (either GET or POST). POST is recommended when using
+     * setTokenAuth() to prevent the token from being recorded in server logs.
```
It might be good to add a comment:

```
When using POST, make sure there's no redirect happening on your server as otherwise POSTed values get lost.
```

silvershadowcc replied:
added commit

silvershadowcc commented:
> 
> 
> Thanks @silvershadowcc
> 
> btw there will be likely still few more changes needed so that the token actually gets posted and not sent through a URL query parameter.

I think it's easiest to parse the URL parameters to JSON and populate $data in sendRequest() ?

silvershadowcc commented:
> 
> 
> Thanks @silvershadowcc
> 
> btw there will be likely still few more changes needed so that the token actually gets posted and not sent through a URL query parameter.

Sorry, I assumed that the parameters would actually be posted haha. Changed it, this works.
Perhaps the variable/method names should be changed and I don't know if this structure is acceptable?

As stated in the commit:
The request method is now set to POST as boolean, instead of a string, since the method should remain defaulted to GET and there is no need to be able to set it to GET.

URL is spliced in URL and parameters which are now fed to the POST payload for both cURL and stream.

tsteur commented on MatomoTracker.php:
```diff
@@ -1533,6 +1548,15 @@ protected function sendRequest($url, $method = 'GET', $data = null, $force = fal
 
+        if ($this->requestMethodPost && !$this->doBulkRequests) {
+            $url_parts = explode('?', $url);
+			
```
fyi not too important but we usually use camel case, eg `urlParts`

tsteur commented:
The namings are all good @silvershadowcc 👍 

> The request method is now set to POST as boolean, instead of a string, since the method should remain defaulted to GET and there is no need to be able to set it to GET.

I wonder if any one would use any other kind of request methods like `PUT` or `HEAD`. I suppose that's not the case. Might be more flexible to force users to specify eg `POST`.

I'll try to test this in the next days @silvershadowcc  looking only at the code it looks good.

The next step after this would be to see if we can somehow make POST the default for the `token_auth` only and send other URL parameters still via GET. Why? This way it's still possible to reply requests using our [log importer](https://github.com/matomo-org/matomo-log-analytics/). What happens is then that in eg apache or nginx log the requests will be found like 

```
IP ... GET matomo.php?action_name=foo&idsite=1&rec=1 ...
```

and someone else can then replay these actions. Vs with a POST of all parameters we would only see `POST matomo.php`. People use log analytics to replay requests if there was a downtime, if they are performing a long lasting upgrade, or in general always. Eg it's a lot faster to not process any of the tracking requests in real time but rather once a day  or a few times a day reply all tracking requests from the logs at once. It's not too important but wanted to provide a bit of background. We don't have to do this as part of this PR but as a next step as part of #43 .
 
We would already POST the token by default. However, we can do this only once we make sure redirected URLs get the value posted as well.

tsteur commented on MatomoTracker.php:
```diff
@@ -1477,6 +1479,19 @@ public function setRequestTimeout($timeout)
+	/**
+     * Sets the request method to POST, which is recommended when using setTokenAuth()
+     * to prevent the token from being recorded in server logs. Avoid using redirects
+     * when using POST to prevent the loss of POST values.
```
Sorry, one more note... can you also add something like

```
When using POST requests, then you won't be able to replay these requests log analytics
```

silvershadowcc commented:
Went back to the previous selection of POST as string. Currently no other methods would be supported since they may require different input for cURL and stream. Converted snake to camel casing and added the warning. Thanks for the background.

tsteur commented on MatomoTracker.php:
```diff
@@ -1477,6 +1477,21 @@ public function setRequestTimeout($timeout)
+     * @param string $method
+     * @return $this
+     */
+    public function setRequestMethod($method)
```
@silvershadowcc could we rename this method to `setRequestMethodForNonBulkRequests` since it doesn't apply to bulk requests? If there's some better name feel free to also use a different name.

silvershadowcc replied:
`setRequestMethodNonBulk` for the sake of brevity? (since it already states "request")

tsteur replied:
Sounds great 👍

silvershadowcc commented:
> If after merging this you are keen on giving #72 a try feel free to go ahead.
Tomorrow :) But I can't test bulk. I added that only POST and GET are supported for docs in the latest commit.

tsteur commented:
Thanks @silvershadowcc
